### DISCUSSION TITLE
Dockview inside shadowdom

### DISCUSSION THREAD

**Discussion by husayt** [mathuo/dockview#424] [Q&A] [answered]
I had tried to run dockview (or even `GridviewComponent` from `dockview-core`) inside shadow-dom and grid constantly sets itself as 0 width and height. I am wondering if it is something i am missing, or grid is not supposed to run inside shadowDom

**Comment by mathuo** [comment]
To be honest, I've never tried it - but I will try it and get back to you. If it doesn't work I would assume it's something trival the library is missing to allow it to work.

**Comment by mathuo** [comment]
I've created a quick CodeSandbox [here](https://codesandbox.io/p/sandbox/shadowdom-example-2h4cps?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clqxyzjgq0006356j8evcmugp%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clqxyzjgq0002356js9xus7ll%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clqxyzjgq0003356jvv497ods%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clqxyzjgq0005356j9gpr2tnd%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clqxyzjgq0002356js9xus7ll%2522%253A%257B%2522id%2522%253A%2522clqxyzjgq0002356js9xus7ll%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqxyzjgq0001356jnleia87t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A16%252C%2522startColumn%2522%253A7%252C%2522endLineNumber%2522%253A16%252C%2522endColumn%2522%253A7%257D%255D%257D%255D%252C%2522activeTabId%2522%253A%2522clqxyzjgq0001356jnleia87t%2522%257D%252C%2522clqxyzjgq0005356j9gpr2tnd%2522%253A%257B%2522id%2522%253A%2522clqxyzjgq0005356j9gpr2tnd%2522%252C%2522activeTabId%2522%253A%2522clqxyzjgq0004356jb21d1qkz%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clqxyzjgq0004356jb21d1qkz%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clr6rrvcc0123356jrf25xml0%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clqxyzjgq0003356jvv497ods%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clqxyzjgq0003356jvv497ods%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) rendering a `DockviewReact` component within a shadow DOM using the `dockview` library.

My very quick example seems to render dimensions ok, are you able to adjust this sandbox to make the dimensions not work as you see?

**Comment by husayt** [comment]
@mathuo this looks really promising. I have been trying to do with GridViewComponent without React.
Here is non shadow dom example.
https://codesandbox.io/p/sandbox/romantic-leakey-f9pfp6?file=%2Fsrc%2Fapp.ts%3A79%2C23

Couldn't get it working inside shadowDom. Is there something special that `GridViewComponent` does, which DockView doesn't?

**Reply by husayt** [reply]
@mathuo here is the the same example but with ShadowDom implemented as Lit webcomponent. As you can see it doesn't work. Am I missing something?

https://codesandbox.io/p/devbox/shadowgridview-9lf936

**Comment by mathuo** [comment, marked as answer]
@husayt I have a feeling [this](https://github.com/mathuo/dockview/blob/master/packages/dockview-core/src/resizable.ts#L48) is the offending line of code... explicitly checks for "within document.body" which would fail for shadow DOM elements.

My example works I assume because its first size by chance is non-zero whereas yours is probably zero (just depends in what order you attach things to the DOM etc I guess - either way it's a bug).

Raised [this](https://github.com/mathuo/dockview/pull/426) PR to fix the issue and have deployed an experimental build `0.0.0-experimental-fd25fdd-20240109` on [npm](https://www.npmjs.com/package/dockview/v/0.0.0-experimental-fd25fdd-20240109) which can be installed through `npm install dockview-core@0.0.0-experimental-fd25fdd-20240109`.

When I use this new version with your example I can see some components do not grow in height as required (outside of the dockview library scope). There are obviously many ways to get the height to work but I just used some `display: flex` and `flex-grow: 1` to achieve this but after that it seems to works nicely - my fork [here](https://codesandbox.io/p/devbox/shadowgridview-forked-66fnkg?file=%2Fpackage.json%3A13%2C19&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clr6txlrr0007356juvf0jmof%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clr6txlrq0002356jlx0q8xx4%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clr6txlrr0004356j9fq4jz8w%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clr6txlrr0006356jspqxpc13%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clr6txlrq0002356jlx0q8xx4%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clr6txlrq0001356jmz1wkdmc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage-lock.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clr6uwtu90002356jj3342zd4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A13%252C%2522startColumn%2522%253A19%252C%2522endLineNumber%2522%253A13%252C%2522endColumn%2522%253A19%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clr6txlrq0002356jlx0q8xx4%2522%252C%2522activeTabId%2522%253A%2522clr6uwtu90002356jj3342zd4%2522%257D%252C%2522clr6txlrr0006356jspqxpc13%2522%253A%257B%2522id%2522%253A%2522clr6txlrr0006356jspqxpc13%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clr6txlrr0005356jxro3zpjj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clr6txlrr0005356jxro3zpjj%2522%257D%252C%2522clr6txlrr0004356j9fq4jz8w%2522%253A%257B%2522id%2522%253A%2522clr6txlrr0004356j9fq4jz8w%2522%252C%2522activeTabId%2522%253A%2522clr6txlrr0003356j1up057rq%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clr6txlrr0003356j1up057rq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clr6uv06l007e356jz78fz5tt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clr6uv05c0014efd3exq92jsh%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

Let me know your thoughts.

Before:
![image](https://github.com/mathuo/dockview/assets/6710312/1a916c24-010f-4ce3-b3b2-39e1eca44c49)

After:
![image](https://github.com/mathuo/dockview/assets/6710312/5aa6011d-75f1-412f-aaeb-6cb5a366f031)

**Reply by husayt** [reply]
This works great. Thank you @mathuo . Need to mention that really impressed with the architecture of this library, having the core done in pure js makes it very easy to build vue, svelte or lit wrappers on top of it. Maybe this example will help to get this started with lit.

**Reply by husayt** [reply]
just noticed this PR. It is certainly the right direction.  I will be very keen to try vue version and help with getting that right

https://github.com/mathuo/dockview/pull/387/files

**Comment by husayt** [comment]
This works as of 1.9.0. Thank you @mathuo
